File: src/main/java/com/base/timeline/Timeline.java
package com.base.timeline;

import com.Global;
import com.base.DateMutableEntity;
import com.base.reference.DMEReference;
import com.base.timeline.change.ChangeID;
import com.base.timeline.change.ChangeSupplier;
import com.base.timeline.change.TimelineChange;
import com.base.timeline.error.SandboxCode;
import com.base.timeline.sandbox.function.SandboxFunctions;
import com.base.timeline.sandbox.core.Objective;
import com.base.timeline.sandbox.core.SandboxHandler;
import com.base.timeline.state.TimelineState;
import com.google.gson.JsonArray;
import com.google.gson.JsonObject;

import javax.annotation.Nullable;
import java.time.LocalDate;
import java.util.*;
import java.util.function.Consumer;

public class Timeline<T extends DateMutableEntity<T>> extends TimelineObject<T> {
    private final TreeMap<LocalDate, TimelineState<T>> timeline = new TreeMap<>();
    //TODO Caching for performance?
    public boolean isLoaded = false;
    public Timeline(T o, DMEReference<T> owner, LocalDate start, @Nullable LocalDate end, List<ChangeSupplier<T,?>> initialState) {
        super(owner);
        //AbstractMutableManager<?,T,?> manager = DMRegistry.getManager(o.getClass());
        timeline.put(start, o.buildBirth(owner,start, initialState));
        if (end == null){
            timeline.put(end, o.buildDeath(owner, Global.MAX_DATE,o.defaultDeathCause(),initialState));
        } else {
            timeline.put(end, o.buildDeath(owner,end,o.defaultDeathCause(),initialState));
        }
        isLoaded = true;
    }

    public Timeline(DMEReference<T> dme){
        super(dme);
    }


    public LocalDate getTimelineMin(){
        return timeline.firstKey();
    }
    public LocalDate getTimelineMax(){
        TimelineState<T> last = timeline.lastEntry().getValue();
        return last.getEnd();
    }
    //Note: Sandbox methods should never be used outside of a sandbox's worker thread! Probably wouldn't break anything, but it's not built for main thread use!
    public void addChange(TimelineChange<T> change){
        //This is the safe way to insert a change using propagation. It should be used by all runtime setters.
        getOrMakeState(change.getStart()); //We just need a state at the exact start date.
        Consumer<SandboxCode> afterChange = (code) -> {
            TimelineState<T> state = getStateAt(change.getStart());
            if (state.isEmpty()){
                removeState(change.getStart());
                TimelineState<T> before = getStateBefore(state.getStart());
                TimelineState<T> after = getStateAfter(state.getStart());
                if (before != null && after != null){
                    before.setEnd(after.getStart().minusDays(1));
                }
            }
        };
        SandboxHandler.StartSandbox(new Objective<>(owner,Global.TimeDirection.FORWARD,change,
                new SandboxFunctions.CanAddChange<>()),getEnd(),null,afterChange);
    }
    public boolean isEmpty(){
        return timeline.isEmpty();
    }
    private void removeState(TimelineState<T> state){
        timeline.remove(state.getStart());
    }
    private void removeState(LocalDate start){
        timeline.remove(start);
    }
    @Override
    public LocalDate getStart(){
        return timeline.firstKey();
    }
    @Override
    public LocalDate getEnd(){
        return timeline.lastKey();
    }

    public void moveStart(LocalDate date){
        final TimelineState<T> startState = getStateAt(getStart());
        moveState(startState,date,null);
    }
    public void moveEnd(LocalDate date){
        final TimelineState<T> endState = getStateAt(getEnd());
        moveState(endState,null,date);
    }


    public boolean isLoaded() {
        return isLoaded;
    }
    public void doTimeChange(LocalDate date){
        final TimelineState<T> t = getStateAt(date);
        for(TimelineChange<? super T> tc : t.getAllChanges(false)){
            tc.apply(getOwner(),t);
        }
    }
    public TimelineState<T> getStateNullable(LocalDate date){
        return timeline.floorEntry(date).getValue();
    }
    public TimelineState<T> getNextState(LocalDate date){
        return timeline.higherEntry(date).getValue();
    }
    public TimelineState<T> getStateAt(LocalDate date){
        TimelineState<T> state = timeline.floorEntry(date).getValue();
        if (state == null){
            throw new IllegalArgumentException("No state found at " + date);
        }
        return state;
    }
    public TimelineState<T> getStateAtExact(LocalDate date, boolean throwIfNotFound){
        TimelineState<T> state = timeline.get(date);
        if (state == null && throwIfNotFound){
            throw new IllegalArgumentException("No state found at " + date);
        }
        return state;
    }
    public TimelineState<T> getStateBefore(LocalDate date){
        TimelineState<T> state = timeline.floorEntry(date.minusDays(1)).getValue();
        date = state.getStart().minusDays(1);
        return timeline.floorEntry(date).getValue();
    }
    public TimelineState<T> getStateAfter(LocalDate date){
        TimelineState<T> state = timeline.ceilingEntry(date.plusDays(1)).getValue();
        date = state.getStart().minusDays(1);
        return timeline.floorEntry(date).getValue();
    }
    public TimelineState<T> getOrMakeState(LocalDate d){
        TimelineState<T> state = timeline.get(d);
        if (state == null){
            return makeNewState(d);
        }
        return state;
    }
    public TimelineState<T>[] getStates(){
        return timeline.values().toArray(TimelineState[]::new);
    }
    public TimelineState<T> getLastState(){
        return timeline.lastEntry().getValue();
    }

    public TimelineState<T> makeNewState(LocalDate start){
        TimelineState<T> before = timeline.floorEntry(start.minusDays(1)).getValue();
        TimelineState<T> after = timeline.ceilingEntry(start.plusDays(1)).getValue();
        if (before.getStart().equals(start)){
            return before;
        } else if(after.getStart().equals(start)) {
            return after;
        }
        if (before.isDuring(start)){
            before.setEnd(start.minusDays(1));
        }
        TimelineState<T> newState = new TimelineState<>(this,start,after.getStart().minusDays(1),false,before.buildFullBreadcrumbList(),new HashMap<>());
        timeline.put(start,newState);
        return newState;
    }

    public boolean moveState(TimelineState<T> state, @Nullable LocalDate newStart, @Nullable LocalDate newEnd){
        boolean changeStart = true;
        boolean changeEnd = true;
        if (newStart == null){
            newStart = state.getStart();
            changeStart = false;
        }
        if (newEnd == null){
            newEnd = state.getEnd();
            changeEnd = false;
        }
        if (!newStart.isBefore(newEnd)){
            logger.error("Cannot move state to a start date that is after the end date! Start: {}, End:{}, State: {}",newStart,newEnd,state);
            return false;
        }
        TimelineState<T> atPlanned = timeline.get(newStart);
        if (atPlanned != null && !atPlanned.equals(state)){
            logger.error("Cannot move state to a start date that is already occupied! Start: {}, End:{}, State: {}",newStart,newEnd,state);
            return false;
        }
        TimelineState<T> beforeEnd = timeline.floorEntry(newEnd.minusDays(1)).getValue();
        TimelineState<T> afterStart = timeline.ceilingEntry(newStart.plusDays(1)).getValue();
        if (afterStart != beforeEnd){
            logger.error("Cannot move state to span over multiple states! Start: {}, End:{}, State: {}",newStart,newEnd,state);
            return false;
        }
        TimelineState<T> afterEnd = timeline.ceilingEntry(newEnd.plusDays(1)).getValue();
        TimelineState<T> beforeStart = timeline.floorEntry(newStart.minusDays(1)).getValue();
        if (changeStart){
            timeline.remove(state.getStart());
            if (beforeStart != null){
                beforeStart.setEnd(newStart.minusDays(1));
            }
            state.setStart(newStart);
            timeline.put(newStart,state);
        }
        if (changeEnd){
            if (afterEnd != null){
                afterEnd.setStart(newEnd.plusDays(1));
            }
            state.setEnd(newEnd);
        }
        return true;
    }
    public void replaceTimeline(JsonObject o){
        isLoaded = false;
        timeline.clear();
        fromJson(o);
        isLoaded = true;
    }
    //==== Shortcut Methods ====
    public final <TC extends TimelineChange<? super T>> TC followBreadcrumb(ChangeID id){
        return followBreadcrumb(this,id);
    }
    public final <TC extends TimelineChange<? super T>> List<TC> findChangeByClassID(LocalDate starting, Global.TimeDirection direction, String classType,
                                                                     final boolean includeDeactivated) {
        return findChangeByClassID(this, starting,direction,includeDeactivated,ChangeID.buildChangeClassID(classType));
    }
    public final <TC extends TimelineChange<? super T>> List<TC> findChangeByClassID(LocalDate starting, Global.TimeDirection direction, TC example,
                                                                                     final boolean includeDeactivated) {
        return findChangeByClassID(this,starting,direction,includeDeactivated,ChangeID.buildChangeClassID(example.getClass().getName()));
    }
    //==== Serializers ====

    public JsonObject toJson(){
        JsonObject o = new JsonObject();
        //Seems wasteful to nest the array in the object, but I'm future proofing here incase TL gets some more variables :)
        JsonArray states = new JsonArray();
        for (Map.Entry<LocalDate, TimelineState<T>> entry : timeline.entrySet()) {
            states.add(entry.getValue().serialize());
        }
        o.add("states",states);
        return o;
    }
    public void fromJson(JsonObject o){
        JsonArray states = o.getAsJsonArray("states");
        for (int i = 0; i < states.size(); i++) {
            TimelineState<T> state = TimelineState.deserialize(this,states.get(i).getAsJsonObject());
            timeline.put(state.getStart(),state);
        }
    }

}

File: src/main/java/com/base/timeline/Timeline.kt
package com.base.timeline

import com.Global
import com.base.AbstractMutableManager
import com.base.DMRegistry
import com.base.DateMutableEntity
import com.base.reference.DMEReference
import com.base.timeline.change.TimelineChange
import com.base.timeline.sandbox.core.Objective
import com.base.timeline.sandbox.core.SandboxHandler
import com.google.gson.JsonArray
import com.google.gson.JsonObject
import java.time.LocalDate
import java.util.*

class Timeline<T : DateMutableEntity<T>> {

    private val timeline: TreeMap<LocalDate, TimelineState<T>> = TreeMap()
    var isLoaded = false
        private set
    private val owner: DMEReference<T>

    constructor(o: T, owner: DMEReference<T>, start: LocalDate, end: LocalDate?, initialState: List<TimelineChange<in T>>) {
        val manager = o.getManager()
        this.owner = owner
        timeline[start] = manager.buildBirth(owner, start, initialState)
        timeline[end ?: Global.MAX_DATE] = manager.buildDeath(owner, end ?: Global.MAX_DATE, initialState)
        isLoaded = true
    }

    constructor(dme: DMEReference<T>) {
        owner = dme
    }

    fun unsafeInsertState(state: TimelineState<T>) {
        insertState(state)
    }

    fun getTimelineMin(): LocalDate = timeline.firstKey()

    fun getTimelineMax(): LocalDate = timeline.lastEntry().value.end

    fun addChange(change: TimelineChange<T>) {
        getOrMakeState(change) // We just need a state at the exact start date.
        SandboxHandler.SandboxApplyChange(Objective(owner, change), change.start, null)
    }

    fun isEmpty(): Boolean = timeline.isEmpty()

    private fun insertState(state: TimelineState<T>) {
        timeline[state.start] = state
    }

    private fun removeState(state: TimelineState<T>) {
        timeline.remove(state.start)
    }

    private fun removeState(start: LocalDate) {
        timeline.remove(start)
    }

    fun getEarliestDate(): LocalDate = timeline.firstKey()

    fun getLatestDate(): LocalDate = timeline.lastKey()

    fun isLast(date: LocalDate): Boolean {
        return timeline.lastEntry().value.start.isBefore(date)
    }

    fun moveBirth(ref: T, date: LocalDate) {
        val oldBirthDate = getEarliestDate()
        val template = DMRegistry.getManager(owner.type)

        if (oldBirthDate.isAfter(date)) {
            val state = getStateAt(oldBirthDate)
            timeline.remove(oldBirthDate)
            timeline[date] = template.buildBirth(owner, date, state.changes(true))
        } else {
            val bt = template.getBirthChange(owner, date)
            SandboxHandler.SandboxApplyChange(Objective(owner, bt), oldBirthDate, null)
            // TODO Sandbox out moving the birth later
        }
    }

    fun moveDeath(ref: T, date: LocalDate) {
        val oldDeathDate = getLatestDate()
        val template = DMRegistry.getManager(owner.type)

        if (oldDeathDate.isBefore(date)) {
            val state = getStateAt(oldDeathDate)
            timeline.remove(oldDeathDate)
            timeline[date] = template.buildDeath(owner, date, state.changes(true))
        } else {
            val bt = template.getDeathChange(owner, date)
            SandboxHandler.SandboxApplyChange(Objective(owner, oldDeathDate, bt), date, null)
        }
    }

    fun doTimeChange(entity: T, date: LocalDate) {
        val t = getStateAt(date)
        t.getChanges(false).forEach { it.apply(entity, t) }
    }

    fun getStateNullable(date: LocalDate): TimelineState<T>? = timeline.floorEntry(date)?.value

    fun getNextState(date: LocalDate): TimelineState<T>? = timeline.higherEntry(date)?.value

    fun getStateAt(date: LocalDate): TimelineState<T> {
        return timeline.floorEntry(date)?.value
            ?: throw IllegalArgumentException("No state found at $date")
    }

    fun getStateAtExact(date: LocalDate, throwIfNotFound: Boolean): TimelineState<T>? {
        val state = timeline[date]
        if (state == null && throwIfNotFound) {
            throw IllegalArgumentException("No state found at $date")
        }
        return state
    }

    fun getStateBefore(date: LocalDate): TimelineState<T> {
        val state = timeline.floorEntry(date.minusDays(1))?.value
        return timeline.floorEntry(state?.start?.minusDays(1))?.value ?: state!!
    }

    fun getStateAfter(date: LocalDate): TimelineState<T> {
        val state = timeline.ceilingEntry(date.plusDays(1))?.value
        return timeline.floorEntry(state?.start?.minusDays(1) ?: date)?.value ?: state!!
    }

    fun getOrMakeState(d: LocalDate): TimelineState<T> {
        return timeline[d] ?: makeNewState(d)
    }

    fun getStates(): Array<TimelineState<T>> {
        return timeline.values.toTypedArray()
    }

    fun getLastState(): TimelineState<T> {
        return timeline.lastEntry().value
    }

    fun save(): JsonObject {
        val obj = JsonObject()
        val states = JsonArray()
        timeline.forEach { (_, value) -> states.add(value.serialize()) }
        obj.add("states", states)
        return obj
    }

    fun load(obj: JsonObject) {
        val states = obj.getAsJsonArray("states")
        for (i in 0 until states.size()) {
            val state = TimelineState.deserialize(states[i].asJsonObject)
            timeline[state.start] = state
        }
    }

    fun makeNewState(start: LocalDate): TimelineState<T> {
        val before = timeline.floorEntry(start.minusDays(1))?.value
        val after = timeline.ceilingEntry(start.plusDays(1))?.value
        if (before?.start == start) return before
        if (after?.start == start) return after

        if (before?.isDuring(start) == true) {
            before.setEnd(start.minusDays(1))
        }

        val breadcrumbs = TimelineHelper.extendTrail(this, before, start)
        val newState = TimelineState(owner, start, after?.start?.minusDays(1), false, breadcrumbs, HashMap())
        timeline[start] = newState
        return newState
    }

    fun replaceTimeline(obj: JsonObject) {
        isLoaded = false
        timeline.clear()
        load(obj)
        isLoaded = true
    }

    fun getOwner(): DMEReference<T> = owner
}